File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.js
function solution(numbers, target) {
    let res = 0;
    
    const dfs = (index, sum) => {
        // 최고 깊이 도달
            if(index === numbers.length) {
                if(sum === target) res++;
            return;
        }
    
        dfs(index + 1, sum + numbers[index]);
        dfs(index + 1, sum - numbers[index]);
    }
    
    dfs(0, 0);
    return res;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
c = 0
temp = []

def add(numbers, signs):
    s = 0
    for i in range(len(numbers)):
        num = numbers[i]
        sign = signs[i]
        
        if sign == '-':
            s -= num
        else:
            s += num
    
    return s

def dfs(numbers, target):
    global c
    if len(temp) == len(numbers):
        s = add(numbers, temp)
        
        if s == target:
            c += 1
        
        return
        
    for i in range(2):
        if i == 0:
            temp.append('-')
        else:
            temp.append('+')
        dfs(numbers, target)
        temp.pop()
    
def solution(numbers, target):
    dfs(numbers, target)
    
    return c
    
        
        
        